### DISCUSSION TITLE
Turn all Pressure BC's off

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#22706] [Q&A General] [answered]
I've got a model with a bunch of different pressures (using Pressure BC) applied to different surfaces. At some time, I'd like to turn all that off (zero pressure everywhere). What's a good way to do that? I've heard about controllable parameters but it doesn't look like any parameters in Pressure BC are controllable?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are the pressure BC directly in the input file or are they added by an Action?
The Control system is able to turn on and off most blocks in the input file at will.
@YaqiWang might have advice on how to do this for an action
If you are refering to this system, https://mooseframework.inl.gov/syntax/BCs/Pressure/index.html
I d investigate the control tags parameter

Another solution is to modify/remove the sidesets when you want to turn the pressure BCs off

Guillaume

**Reply by jessecarterMOOSE** [reply]
I'm using the action if that helps. Actually, I have my own custom action that calls the Pressure action!

**Reply by GiudGiud** [reply]
ok so the control_tags parameter is likely not forwarded at this point down to the BCs.

I would try to add the show_actions parameter to [Debug], get the names of the pressure BCs from there, then specify those names in a ConditionalEnableControl
```
  [turn_off]
    type = ConditionalFunctionEnableControl
    disable_objects = 'BCs::name'
    conditional_function = 'conditional_function'
    execute_on = 'timestep_begin'
  []
```

if this does not work because the object name is not in the input file, then I would look at forwarding the control_tags parameter.
You can look at this PR to see how this was worked on recently, I may be mistaken

https://github.com/idaholab/moose/pull/22069

**Reply by jessecarterMOOSE** [reply]
@GiudGiud sanity check - What's a simple way to get a signal this is working, and working for all Pressure BCs (of which there may be many)? Maybe a Postprocessor I can use to check that the applied force is zero?

**Reply by GiudGiud** [reply]
ah I have a PR that can do that for you but it s not in yet :(

You can put a print statement in the BC
or
you can design a postprocessor to measure the force

**Reply by jessecarterMOOSE** [reply]
@GiudGiud another question: can I add the ConditionalFunctionEnableControl object from an action? I don't see an addControl method.

**Reply by GiudGiud** [reply]
If it s been done it will have been in the thermal hydraulics module

**Reply by GiudGiud** [reply]
THMAddControlAction.C seems to have some logic for that

**Reply by jessecarterMOOSE** [reply]
How much of that logic do I need for a non-THM problem?

**Comment by jessecarterMOOSE** [comment]
@GiudGiud Now that your PR on this is merged, can you explain/demostrate how you would do this *from another action*?. Right now, I have an action that adds PressureAction like 
```
std::string surface_name = "top";
auto params = _action_factory.getValidParams("PressureAction");
params.set<std::vector<BoundaryName>>("boundary") = {surface_name};
params.set<Real>("factor") = factor;
std::string obj_name = surface_name+"_pressure";
auto action = MooseSharedNamespace::static_pointer_cast<MooseObjectAction>(_action_factory.create("PressureAction", obj_name, params));
_awh.addActionBlock(action);
```

Then later on in my Action I add a Control. How do I point it at this?

**Reply by GiudGiud** [reply]
effectively, control the "enable" parameter of the PressureAction.

so either through some enable/disable parameters of control
```
  [pressure_crank]
    type = TimePeriod
    enable_objects = 'BCs/Pressure/Side1'
    disable_objects = 'BCs/Pressure/Side2 BCs/Pressure/Side3'
```

or control the enable boolean for a control that acts on booleans

**Reply by jessecarterMOOSE** [reply]
OK so effectively my disable/enable_objects needs to be `BCs/Pressure/obj_name`?

**Reply by GiudGiud** [reply]
I think so because that s what the action is effectively creating

by obj_name, it s the PressureAction name

**Reply by jessecarterMOOSE** [reply]
hmm, `The desired parameter 'BCs/Pressure/top_pressure/enable' was not located... `

**Reply by jessecarterMOOSE** [reply]
`top_pressure` is my obj_name here

**Reply by GiudGiud** [reply]
can you remove this line of code for now then use the brand new Debug/show_controllable to get the actual name/syntax to use

**Reply by jessecarterMOOSE** [reply]
do you mean remove the check in `Control.C`?

**Reply by GiudGiud** [reply]
no I mean removing any instance of "connectControllable" that you are currently using

are you saying that your action, creating the PressureAction, is erroring right now?

**Reply by jessecarterMOOSE** [reply]
no it's a runtime error - when it comes time for the Control to activate and turn PressureAction off, I get that error

My Control has: `disable_objects = 'BCs/Pressure/top_pressure'`

If I use PressureAction in an input file like this, it seems to work:
```
  [Pressure]
    [top_pressure]
      boundary = top
      factor = -10
      use_displaced_mesh = true
    []
  []
  ```
 But if I add a PressureAction from inside my Action (code shown above) I get that error

**Reply by GiudGiud** [reply]
what does `Debug/show_controllable` say ?

**Reply by jessecarterMOOSE** [reply]
A lot! Under which condition? Pressure action added with input file or muly Action? Maybe I'll do both and compare

**Reply by GiudGiud** [reply]
in your final setup, without the Control so it does not error

it should show you what syntax to use for the Control enable/disable parameter

**Reply by jessecarterMOOSE** [reply]
It works fine if I do everything in an input file (PressureAction and Control). The issue is when I try to add PressureAction in my custom Action instead. I do not understand what it's doing differently. 

When I use just the input file, I see this in the debug output:
```
Controls:
  PressureActionBCs/Pressure/top_pressure
    Name(s): Action::PressureActionBCs/Pressure/top_pressure BCs/Pressure/top_pressure
    Tag(s): BCs/Pressure
    Parameter(s):
      enable                     bool
```

When I add PressureAction in my custom Action, i get:
```
Controls:
  PressureActiontop_pressure
    Name(s): Action::PressureActiontop_pressure
    Parameter(s):
      enable                     bool
```

Notice the `Tag(s)` line is missing in the latter case and also the naming is slightly different.

**Reply by GiudGiud** [reply]
ok I ll try something

**Reply by GiudGiud** [reply]
can you try setting a control_tag parameter in your action, and forwarding it to the PressureAction

**Reply by GiudGiud** [reply]
like this
```
      pressure_action_params.set<std::vector<std::string>>("control_tags") =
          getParam<std::vector<std::string>>("control_tags");
```

or you can hardcode it but probably not necessary

**Reply by jessecarterMOOSE** [reply]
I put this in my action:
```
params.set<std::vector<std::string>>("control_tags") = {"BCs/Pressure"};
```
To try and match what I see when I use PressureAction in input file

Now Debug controls shows:
```
Controls:
  PressureActiontop_pressure
    Name(s): Action::PressureActiontop_pressure BCs/Pressure/PressureActiontop_pressure
    Tag(s): BCs/Pressure
    Parameter(s):
      enable                     bool
```
The Tag(s) line is the same but the Name(s) line is still a bit different

What do I put in my Control? Name or Tag?

**Reply by GiudGiud** [reply]
name should be fine. 
in the control, I think you should have
tag/name/parameter

so here:
```
BCs/Pressure/<name of your action>/enable
```

**Reply by GiudGiud** [reply]
note that this syntax should work too?
```
BCs/Pressure/*/enable
```

**Reply by jessecarterMOOSE** [reply]
I guess I'm not clear on the meaning of Name and Tag and how/where they are used. I also noticed that your `pressure_control_test.i` works if I comment out `control_tags`
